file/file_prefetch_buffer.cc

```diff
@@ -21,90 +20,282 @@
+
+    memcpy(bufs_[2].buffer_.BufferStart(), bufs_[second].buffer_.BufferStart(),
+           length);
+
+    curr_ = curr_ ^ 1;
+  }
+  bufs_[2].offset_ = offset;
+  bufs_[2].buffer_.Size(length);
+}
+
+// In async_read = true:
+// async_read is enabled in case of sequential reads. So when
+// buffers are switched, we clear the curr_ buffer as we assume the data has
+// been consumed because of sequential reads.
+//
+// Scenarios for prefetching asynchronously:
+// Case1: If both buffers are empty, prefetch n bytes
+//        synchronusly in curr_
+//        and prefetch readahead_size_/2 async in second buffer.
+// Case2: If second buffer has partial or full data, make it current and
+//        prefetch readahead_size_/2 async in second buffer. In case of
+//        partial data, prefetch remaining bytes from size n synchronously.
+// Case3: If curr_ has partial data, prefetch remaining bytes from size n
+//        synchronously in curr_ and prefetch readahead_size_/2 bytes async in
+//        second buffer.
+// Case4: [Corner case] If curr1 has some data,
+//        remaining is in second buffer.
+//
+// If async_read = false:
+// all the bytes should be prefetched synchronously in curr_ and second buffer
+// should be empty and never used.
+Status FilePrefetchBuffer::Prefetch(const IOOptions& opts,
+                                    RandomAccessFileReader* reader,
+                                    uint64_t offset, size_t n,
+                                    Env::IOPriority rate_limiter_priority,
+                                    bool is_async_read, bool* third_buffer) {
+  if (!enable_ || reader == nullptr) {
+    return Status::OK();
+  }
+
+  Status s;
+  TEST_SYNC_POINT("FilePrefetchBuffer::Prefetch:Start");
+
+  if (offset + n <= bufs_[curr_].offset_ + bufs_[curr_].buffer_.CurrentSize()) {
+    // All requested bytes are already in the curr_ buffer. So no need to Read
+    // again.
+    return s;
+  }
+
+  // Index of second buffer.
+  uint32_t second = curr_ ^ 1;
+  size_t alignment = reader->file()->GetRequiredBufferAlignment();
+
+  // Case: If data is in second buffer, make it curr_.
+  // Second buffer can be either partial filled or full.
+  if (is_async_read && offset >= bufs_[second].offset_ &&
+      offset <= bufs_[second].offset_ + bufs_[second].buffer_.CurrentSize()) {
+    // Switch the buffers.
+    bufs_[curr_].buffer_.Clear();
+    curr_ = curr_ ^ 1;
+    second = curr_ ^ 1;
+  }
+
+  // If second buffer contains outdated data, clear it. It can be because
+  // sequential reads were read from the cache instead of this buffer.
+  if (bufs_[second].buffer_.CurrentSize() > 0 &&
+      offset >= bufs_[second].offset_ + bufs_[second].buffer_.CurrentSize()) {
+    bufs_[second].buffer_.Clear();
+  }
+
+  // Case: If some of the data is in buffer 1 and rem in buf2.
+  if (is_async_read && third_buffer && offset >= bufs_[curr_].offset_ &&
+      offset < bufs_[curr_].offset_ + bufs_[curr_].buffer_.CurrentSize() &&
```

Comment:
If data is in both buffers, It prefetch synchronously (n+readahead/2). For that it first checks how much data is in curr_ buffer and  if prefetched is required, it will prefetch sync in second buffer.